src/MOI_wrapper.jl

```diff
@@ -0,0 +1,355 @@
+
+"""
+    Manopt.Optimizer()
+
+Creates a new optimizer object.
+
+The minimization of a function `f(X)` of an an array `X[1:n1,1:n2,...]`
+over a manifold `M` starting at `X0`, can be modeled as follows:
+```julia
+using JuMP
+@variable(model, X[i1=1:n1,i2=1:n2,...] in M, start = X0[i1,i2,...])
+@objective(model, Min, f(X))
+```
+"""
+mutable struct Optimizer <: MOI.AbstractOptimizer
+    manifold::Union{Nothing,ManifoldsBase.AbstractManifold}
+    problem::Union{Nothing,AbstractManoptProblem}
+    state::Union{Nothing,AbstractManoptSolverState}
+    variable_primal_start::Vector{Union{Nothing,Float64}}
+    sense::MOI.OptimizationSense
+    nlp_data::MOI.NLPBlockData
+    qp_data::QPBlockData{Float64}
+    options::Dict{String,Any}
+    function Optimizer()
+        return new(
+            nothing,
+            nothing,
+            nothing,
+            Union{Nothing,Float64}[],
+            MOI.FEASIBILITY_SENSE,
+            MOI.NLPBlockData([], _EmptyNLPEvaluator(), false),
+            QPBlockData{Float64}(),
+            Dict{String,Any}(),
+        )
+    end
+end
+
+MOI.get(::Optimizer, ::MOI.SolverVersion) = "0.4.25"
+
+function MOI.is_empty(model::Optimizer)
+    return isnothing(model.manifold) &&
+           isempty(model.variable_primal_start) &&
+           model.nlp_data.evaluator isa _EmptyNLPEvaluator &&
+           model.sense == MOI.FEASIBILITY_SENSE
+end
+
+function MOI.empty!(model::Optimizer)
+    model.manifold = nothing
+    model.problem = nothing
+    model.state = nothing
+    empty!(model.variable_primal_start)
+    model.sense = MOI.FEASIBILITY_SENSE
+    model.nlp_data = MOI.NLPBlockData([], _EmptyNLPEvaluator(), false)
+    model.qp_data = QPBlockData{Float64}()
+    return nothing
+end
+
+function MOI.supports(::Optimizer, ::MOI.RawOptimizerAttribute)
+    # FIXME It should depend on `attr.name`
+    return true
+end
+
+function MOI.get(model::Optimizer, attr::MOI.RawOptimizerAttribute)
+    if !MOI.supports(model, attr)
+        throw(MOI.UnsupportedAttribute(attr))
+    end
+    return model.options[attr.name]
+end
+
+function MOI.set(model::Optimizer, attr::MOI.RawOptimizerAttribute, value)
+    if !MOI.supports(model, attr)
+        throw(MOI.UnsupportedAttribute(attr))
+    end
+    model.options[attr.name] = value
+    return nothing
+end
+
+MOI.get(::Optimizer, ::MOI.SolverName) = "Manopt"
+
+MOI.supports_incremental_interface(::Optimizer) = true
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike)
+    return MOI.Utilities.default_copy_to(dest, src)
+end
+
+function MOI.supports_add_constrained_variables(::Optimizer, ::Type{<:VectorizedManifold})
+    return true
+end
+
+function MOI.add_constrained_variables(model::Optimizer, set::VectorizedManifold)
+    F = MOI.VectorOfVariables
+    if !isnothing(model.manifold)
+        throw(
+            AddConstraintNotAllowed{F,typeof(set)}(
+                "Only one manifold allowed, variables in `$(model.manifold)` have already been added.",
```

Comment:
In principle one could model two manifolds as a `Productmanifold` with the `ArrayPartition`, but that might be a bit technical for a first PR.

Reply:
Since this is a constraint, I'd assume that more than one manifold here means an intersection in the embedding and we don't support that currently. What could be done with not too much work is finding which manifolds are level-set-based, convert them to equality constraints, and (if possible) leave only one as an actual manifold constraint.

Reply:
Yes, it might be a nice addition and not too hard as a follow up PR